src/trashholder.cpp

```diff
@@ -9,40 +9,47 @@
 
 extern Game g_game;
 
-ReturnValue TrashHolder::queryAdd(int32_t, const Thing&, uint32_t, uint32_t, Creature*) const
+ReturnValue TrashHolder::queryAdd(int32_t, std::shared_ptr<const Thing>, uint32_t, uint32_t,
+                                  std::shared_ptr<Creature>) const
 {
 	return RETURNVALUE_NOERROR;
 }
 
-ReturnValue TrashHolder::queryMaxCount(int32_t, const Thing&, uint32_t count, uint32_t& maxQueryCount, uint32_t) const
+ReturnValue TrashHolder::queryMaxCount(int32_t, std::shared_ptr<const Thing>, uint32_t count, uint32_t& maxQueryCount,
+                                       uint32_t) const
 {
 	maxQueryCount = std::max<uint32_t>(1, count);
 	return RETURNVALUE_NOERROR;
 }
 
-ReturnValue TrashHolder::queryRemove(const Thing&, uint32_t, uint32_t, Creature* /*= nullptr*/) const
+ReturnValue TrashHolder::queryRemove(std::shared_ptr<const Thing>, uint32_t, uint32_t,
+                                     std::shared_ptr<Creature> /*= nullptr*/) const
 {
 	return RETURNVALUE_NOTPOSSIBLE;
 }
 
-Cylinder* TrashHolder::queryDestination(int32_t&, const Thing&, Item**, uint32_t&) { return this; }
+std::shared_ptr<Thing> TrashHolder::queryDestination(int32_t&, std::shared_ptr<const Thing>, std::shared_ptr<Item>&,
+                                                     uint32_t&)
+{
+	return shared_from_this();
+}
 
-void TrashHolder::addThing(Thing* thing) { return addThing(0, thing); }
+void TrashHolder::addThing(std::shared_ptr<Thing> thing) { return addThing(0, thing); }
 
-void TrashHolder::addThing(int32_t, Thing* thing)
+void TrashHolder::addThing(int32_t, std::shared_ptr<Thing> thing)
 {
-	Item* item = thing->getItem();
+	auto item = thing->getItem();
 	if (!item) {
 		return;
 	}
 
-	if (item == this || !item->hasProperty(CONST_PROP_MOVEABLE)) {
+	if (item.get() == this || !item->hasProperty(CONST_PROP_MOVEABLE)) {
 		return;
 	}
 
 	const ItemType& it = Item::items[id];
 	if (item->isHangable() && it.isGroundTile()) {
-		Tile* tile = dynamic_cast<Tile*>(getParent());
+		auto tile = Item::getTile();
```

Comment:
Could this be `const auto`
```suggestion
		cont auto tile = Item::getTile();
```